extensions/context/src/main/java/io/opentelemetry/context/Context.java

```diff
@@ -0,0 +1,214 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * A context propagation mechanism which can carry scoped-values across API boundaries and between
+ * threads.
+ *
+ * <p>A Context object can be {@linkplain #attach attached} to the {@link ContextStorage}, which
+ * effectively forms a <b>scope</b> for the context. The scope is bound to the current thread.
+ * Within a scope, its Context is accessible even across API boundaries, through {@link #current}.
+ * The scope is later exited by {@link Scope#close()} closing} the scope.
+ *
+ * <p>Context objects are immutable and inherit state from their parent. To add or overwrite the
+ * current state a new context object must be created and then attached, replacing the previously
+ * bound context. For example:
+ *
+ * <pre>{@code
+ * Context withCredential = Context.current().withValue(CRED_KEY, cred);
+ * withCredential.wrap(new Runnable() {
+ *   public void run() {
+ *      readUserRecords(userId, CRED_KEY.get());
+ *   }
+ * }).run();
+ * }</pre>
+ *
+ * <p>Notes and cautions on use:
+ *
+ * <ul>
+ *   <li>Every {@code attach()} must be followed by a {@code Scope#close()}. Breaking these rules
+ *       may lead to memory leaks.
+ *   <li>While Context objects are immutable they do not place such a restriction on the state they
+ *       store.
+ *   <li>Context is not intended for passing optional parameters to an API and developers should
+ *       take care to avoid excessive dependence on context when designing an API.
+ *   <li>Attaching Context from a different ancestor will cause information in the current Context
+ *       to be lost. This should generally be avoided.
+ * </ul>
+ */
+public interface Context {
+
+  /** Return the context associated with the current scope. */
+  static Context current() {
+    Context current = ContextStorage.get().current();
+    return current != null ? current : root();
+  }
+
+  /**
+   * Returns the root {@link Context} which all other {@link DefaultContext} are derived from.
+   *
+   * <p>It should generally not be required to use the root {@link Context} directly - instead, use
+   * {@link Context#current()} to operate on the current {@link Context}. Only use this method if
+   * you are absolutely sure you need to disregard the current {@link Context} - this almost always
+   * is only a workaround hiding an underlying context propagation issue.
+   */
+  static Context root() {
+    return DefaultContext.ROOT;
+  }
+
+  /**
+   * Returns the value stored in this {@link Context} for the given {@link DefaultContextKey}, or
+   * {@code null} if there is no value for the key in this context.
+   */
+  @Nullable
+  <V> V getValue(ContextKey<V> key);
+
+  /**
+   * Returns a new context with the given key value set.
+   *
+   * <pre>{@code
+   * Context withCredential = Context.current().withValue(CRED_KEY, cred);
+   * withCredential.run(new Runnable() {
```

Comment:
I think this needs to change to be "wrap"

Reply:
@anuraaga ?

Reply:
@jkwatson I think this is addressed, maybe you found an older commit?